File: app/src/main/java/com/github/shadowsocks/QuickToggleShortcut.java
package com.github.shadowsocks;


import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ShortcutManager;
import android.os.Build;
import android.os.Bundle;
import android.os.RemoteException;

import androidx.annotation.Nullable;

import com.github.shadowsocks.utils.Constants;
import com.github.shadowsocks.utils.ToastUtils;
import com.github.shadowsocks.utils.Utils;

/**
 * @author Mygod
 */
public class QuickToggleShortcut extends Activity {

    private ServiceBoundContext mServiceBoundContext;

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(newBase);
        mServiceBoundContext = new ServiceBoundContext(newBase) {
            @Override
            protected void onServiceConnected() {
                try {
                    int state = bgService.getState();
                    switch (state) {
                        case Constants.State.STOPPED:
                            ToastUtils.showShort(R.string.loading);
                            Utils.INSTANCE.startSsService(this);
                            break;
                        case Constants.State.CONNECTED:
                            Utils.INSTANCE.stopSsService(this);
                            break;
                        default:
                            // ignore
                            break;
                    }
                } catch (RemoteException e) {
                    e.printStackTrace();
                }
                finish();
            }
        };
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        String action = getIntent().getAction();

        if (Intent.ACTION_CREATE_SHORTCUT.equals(action)) {
            setResult(Activity.RESULT_OK, new Intent()
                    .putExtra(Intent.EXTRA_SHORTCUT_INTENT, new Intent(this, QuickToggleShortcut.class))
                    .putExtra(Intent.EXTRA_SHORTCUT_NAME, getString(R.string.quick_toggle))
                    .putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE,
                            Intent.ShortcutIconResource.fromContext(this, R.mipmap.ic_launcher)));
            finish();
        } else {
            mServiceBoundContext.attachService();
            if (Build.VERSION.SDK_INT >= 25) {
                ShortcutManager service = getSystemService(ShortcutManager.class);
                if (service != null) {
                    service.reportShortcutUsed("toggle");
                }
            }
        }
    }

    @Override
    protected void onDestroy() {
        mServiceBoundContext.detachService();
        super.onDestroy();
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/QuickToggleShortcut.kt
package com.bige0.shadowsocksr

import android.app.*
import android.content.*
import android.content.pm.*
import android.os.*
import androidx.core.content.*
import androidx.core.content.pm.*
import com.bige0.shadowsocksr.shortcuts.*
import com.bige0.shadowsocksr.utils.*

class QuickToggleShortcut : Activity()
{
	private lateinit var mServiceBoundContext: ServiceBoundContext

	override fun attachBaseContext(newBase: Context)
	{
		super.attachBaseContext(newBase)
		mServiceBoundContext = object : ServiceBoundContext(newBase)
		{
			override fun onServiceConnected()
			{
				try
				{
					when (bgService?.state)
					{
						Constants.State.STOPPED ->
						{
							ToastUtils.showShort(R.string.loading)
							Utils.startSsService(this)
							val connectedShortcut = makeToggleShortcut(context = this, isConnected = true)
							ShortcutManagerCompat.pushDynamicShortcut(this, connectedShortcut)
						}

						Constants.State.CONNECTED ->
						{
							Utils.stopSsService(this)
							val disconnectedShortcut = makeToggleShortcut(context = this, isConnected = false)
							ShortcutManagerCompat.pushDynamicShortcut(this, disconnectedShortcut)
						}

						else ->
						{
						}
					} // ignore
				}
				catch (e: RemoteException)
				{
					e.printStackTrace()
				}

				finish()
			}
		}
	}

	override fun onCreate(savedInstanceState: Bundle?)
	{
		super.onCreate(savedInstanceState)

		if (intent.action == Intent.ACTION_MAIN)
		{
			mServiceBoundContext.attachService()
			if (Build.VERSION.SDK_INT >= 25)
			{
				getSystemService<ShortcutManager>()!!.reportShortcutUsed("toggle")
			}
		}
	}

	override fun onDestroy()
	{
		mServiceBoundContext.detachService()
		super.onDestroy()
	}
}
